Makefile

```diff
@@ -36,111 +33,214 @@ TESTSRC := $(COMMON_TESTSRC) $(FUNCTIONAL_TESTSRC) $(UNIT_TESTSRC)
+
+
+###################################################################
+# BUILD DIRECTORIES AND FILES
+#
 
-debug-log: CFLAGS = -g -DDEBUG -DCC_LOG $(DEFAULT_CFLAGS)
-debug-log: LDFLAGS = -g $(DEFAULT_LDFLAGS)
-debug-log: .debug-log all
-	rm -f .release .debug
+OBJDIR = $(BUILD_DIR)/obj
+BINDIR = $(BUILD_DIR)/bin
+LIBDIR = $(BUILD_DIR)/lib
 
-.release:
-	$(MAKE) clean
-	touch .release
+OBJ := $(SRC:%.c=$(OBJDIR)/%.o)
 
-.debug:
-	$(MAKE) clean
-	touch .debug
+# Objects from test sources in tests/ that are shared by functional/ and unit/ tests
+COMMON_TESTOBJ= $(COMMON_TESTSRC:%.c=$(OBJDIR)/%.o)
 
-.debug-log:
-	$(MAKE) clean
-	touch .debug-log
+# Objects from test sources in tests/functional/ sub-dir
+FUNCTIONAL_TESTOBJ= $(FUNCTIONAL_TESTSRC:%.c=$(OBJDIR)/%.o)
+
+# Objects from unit-test sources in tests/unit/ sub-dir, for fast unit-tests
+# Resolves to a list: obj/tests/unit/a.o obj/tests/unit/b.o obj/tests/unit/c.o
+FAST_UNIT_TESTOBJS= $(FAST_UNIT_TESTSRC:%.c=$(OBJDIR)/%.o)
+
+# ----
+# Binaries from unit-test sources in tests/unit/ sub-dir
+# Although the sources are in, say, tests/unit/splinterdb_quick_test.c, and so on
+# the binaries are named bin/unit/splinterdb_quick_test.
+# Also, there may be other shared .c files that don't yield a standalone
+# binary. Hence, only build a list from files named *_test.c
+# Resolves to a list: bin/unit/a_test bin/unit/b_test bin/unit/c_test ...
+UNIT_TESTBIN_SRC=$(filter %_test.c, $(UNIT_TESTSRC))
+UNIT_TESTBINS=$(UNIT_TESTBIN_SRC:$(TESTS_DIR)/%_test.c=$(BINDIR)/%_test)
+
+####################################################################
+# The main targets
+#
+
+all: libs all-tests $(EXTRA_TARGETS)
+libs: $(LIBDIR)/libsplinterdb.so $(LIBDIR)/libsplinterdb.a
+all-tests: $(BINDIR)/driver_test $(BINDIR)/unit_test $(UNIT_TESTBINS)
+
+
+# This is for backwards compatibility with old CI.  Once we update CI,
+# we can delete this target
+debug: all
+
+
+#######################################################################
+# CONFIGURATION CHECKING
+#
+# Save a hash of the config we used to perform the build and check for
+# any mismatched config from a prior build, so we can ensure we never
+# accidentially build using a mixture of configs
+
+CONFIG_HASH = $(shell echo $(CC) $(DEPFLAGS) $(CFLAGS) $(INCLUDE) $(TARGET_ARCH) $(LD) $(LDFLAGS) $(LIBS) $(AR) | md5sum | cut -f1 -d" ")
+CONFIG_FILE_PREFIX = $(BUILD_DIR)/build-config.
+CONFIG_FILE = $(CONFIG_FILE_PREFIX)$(CONFIG_HASH)
+
+.PHONY: mismatched_config_file_check
+mismatched_config_file_check: | $(BUILD_DIR)/.
+	$(BRIEF_PARTIAL) Checking for mismatched config...
+	$(COMMAND) ls $(CONFIG_FILE_PREFIX)* 2>/dev/null | grep -v $(CONFIG_FILE) | xargs -ri sh -c 'echo "Mismatched config file \"{}\" detected.  You need to \"make clean\"."; false'
+	$(BRIEF) No mismatched config found
+
+
+$(CONFIG_FILE): | $(BUILD_DIR)/. mismatched_config_file_check
+	$(BRIEF) Saving config to $@
+	$(COMMAND) echo CC          = $(CC)          >> $@
```

Comment:
As these `Makefile` symbols are essentially derived from  any existing `BUILD_*` env-var settings, can you also consider dumping the following to the config check file:

`env | grep BUILD_ >> $@`

Add this to before L235.